### ISSUE TITLE
Utilize Baggage Claim as a CSI Driver

### ISSUE THREAD

**Issue by scottietremendous** [concourse/concourse#6591]
#### TLDR

- We've earmarked this as the first epic in the larger track of work around implementing a [native kubernetes](https://github.com/concourse/concourse/issues/5682) runtime. After spiking on the [CSI driver interface](https://github.com/concourse/concourse/issues/6133), we've decided to create a CSI driver based on baggageclaim. 

#### Problem

- As we started spiking about the potential of a k8s runtime, we realize that we needed to figure out what our storage solution would be before iterating on any other features. 
- Storage has a large effect on how Concourse interacts with the runtime. 
- In a container orchestration (CO) system such as kubernetes, you need a way to provide storage to containers. In order to support many different storage providers, the CSI spec was made to allow CO's to have a consistent way to ask for storage containers. 
- Baggageclaim is a tool Concourse uses to create volumes and mounts/places them to later create a container, but doesn't understand the CSI interface. 
- Baggageclaim will need some persistent disk, so our CSI driver solution depends on k8s having another CSI driver that can give baggageclaim an HD to store all the volumes it makes on. 

#### Proposed Solution

- Implement Baggageclaim as a CSI Driver
- Follow the [recommended mechanism](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md#recommended-mechanism-for-deploying-csi-drivers-on-kubernetes) for deploying CSI Drivers on Kubernetes, with a few small variations:  
   - no `external-resizer` container. Not planning to support resizing.
   - no `external-snapshotter` container. The baggageclaim CSI driver will use the `CLONE_VOLUME` feature to create COW volumes in baggageclaim instead of trying to use snapshots.
   - An extra volume must be mounted for each replica Pod in the DaemonSet. This volume, which should be very large, will be used by baggageclaim to store the volumes that it creates on each Kubernetes node.
   - The baggageclaim CSI drver will not guarantee the requested storage capicity because Concourse has no idea how much space any given step in will use, therefore Concourse continues to expect the CSI driver to follow this assumption. Kubernetes will force us to specify a storage request but the baggageclaim CSI driver will ignore this value. This goes against the CSI spec.
 
#### Outcomes

- We don't anticipate any large performance leaps with this implementation, but this work will go a long way in driving out the rest of the k8s runtime track.

#### Current Status

- Will be picked up for engineering following the completion of the [`containerd`: leverage privileged containers](https://github.com/concourse/concourse/issues/5713) epic

#### References

- [k8s storage RFC](https://github.com/concourse/rfcs/blob/074-k8s-storage/074-k8s-storage/proposal.md)

**Comment by multimac** [comment]
I've begun work and have a POC for a Baggage Claim CSI driver in this repository...

https://github.com/multimac/concourse-k8s-worker

It's fairly basic, in that it reuses the existing overlay driver to manage the volumes and propagates the mounts into the Pods when requested.

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
